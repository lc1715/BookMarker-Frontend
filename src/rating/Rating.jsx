import { useContext } from 'react';
import UserContext from '../auth/UserContext';
import Box from '@mui/material/Box';


function Rating({ rating, setRating, addRating, updateRating, hover, setHover }) {
    const { currentUser } = useContext(UserContext);

    async function submitRating(star) {
        console.log('star=', star)

        if (!rating) {
            await addRating(star);
        } else {
            await updateRating(star)
        }

        setRating(star);
    };

    function loggedInRating() {
        return (
            <div>
                {[1, 2, 3, 4, 5].map((star) => (
                    <Box component="span"
                        sx={{
                            color: star <= (hover || rating) ? '#e87400' : 'gray',
                            cursor: 'pointer',
                            fontSize: '38px'
                        }}
                        onClick={() => submitRating(star)}
                        onMouseEnter={() => setHover(star)}
                        onMouseLeave={() => setHover(null)}
                        key={star}
                    >
                        {' '}★{' '}
                    </Box>
                ))}
            </div>
        )
    };

    function loggedOutRating() {
        return (
            <div>
                {[1, 2, 3, 4, 5].map((star) => (
                    <Box component="span"
                        sx={{
                            color: 'gray',
                            cursor: 'pointer',
                            fontSize: '38px'
                        }}
                        onClick={() => submitRating(star)}
                        key={star}
                    >
                        {' '}★{' '}
                    </Box>
                ))}
            </div>
        )
    };

    return (
        <>
            {currentUser ? loggedInRating() : loggedOutRating()}
        </>
    )
};

export default Rating;

