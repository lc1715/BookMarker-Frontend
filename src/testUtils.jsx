import UserContext from "./auth/UserContext";

const demoUser = {
    id: 1,
    username: 'testUser',
    email: 'test@email.com',
    volume_ids: ['e3', 't5'],
}

const UserProvider =
    ({ children, currentUser = demoUser, hasSavedBook = () => (true) }) => (
        <UserContext.Provider value={{ currentUser, hasSavedBook }}>
            {children}
        </UserContext.Provider>
    );

export { UserProvider };
