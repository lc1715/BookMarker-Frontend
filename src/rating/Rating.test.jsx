import { expect, it } from 'vitest';
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from '../testUtils';
import Rating from './Rating';

it('renders Rating component', () => {
    render(
        <MemoryRouter>
            <UserProvider>
                <Rating />
            </UserProvider>
        </MemoryRouter>
    )
})

it('matches with snapshot', async () => {
    const { asFragment } = render(
        <MemoryRouter>
            <UserProvider>
                <Rating />
            </UserProvider>
        </MemoryRouter>
    )

    expect(asFragment()).toMatchSnapshot();
})

it('has the correct rating data', async () => {
    render(
        <MemoryRouter>
            <UserProvider>
                <Rating />
            </UserProvider>
        </MemoryRouter>
    )

    expect(screen.getAllByText('â˜…')).toHaveLength(5);
});
