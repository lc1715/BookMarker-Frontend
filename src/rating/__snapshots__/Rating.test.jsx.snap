// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches with snapshot 1`] = `
<DocumentFragment>
  <div>
    <span
      class="MuiBox-root css-1h0szju"
    >
       ★ 
    </span>
    <span
      class="MuiBox-root css-1h0szju"
    >
       ★ 
    </span>
    <span
      class="MuiBox-root css-1h0szju"
    >
       ★ 
    </span>
    <span
      class="MuiBox-root css-1h0szju"
    >
       ★ 
    </span>
    <span
      class="MuiBox-root css-1h0szju"
    >
       ★ 
    </span>
  </div>
</DocumentFragment>
`;
